File: scripts/verify_buildkit_attestations.py
#!/usr/bin/env python3
"""Fail closed unless BuildKit published complete SBOM and SLSA in-toto blobs."""

import argparse
import hashlib
import json
import pathlib
import subprocess
from typing import Any

IN_TOTO_MEDIA_TYPE = "application/vnd.in-toto+json"
ATTESTATION_ARTIFACT_TYPE = "application/vnd.docker.attestation.manifest.v1+json"
SBOM_PREDICATE = "https://spdx.dev/Document"
SLSA_PREDICATES = {
    "https://slsa.dev/provenance/v0.2",
    "https://slsa.dev/provenance/v1",
}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--crane", required=True)
    parser.add_argument("--image", required=True)
    parser.add_argument("--digest", required=True)
    parser.add_argument("--output-dir", required=True, type=pathlib.Path)
    return parser.parse_args()


def crane_bytes(crane: str, *arguments: str) -> bytes:
    result = subprocess.run(
        [crane, *arguments],
        check=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    return result.stdout


def parse_verified_json(body: bytes, digest: str) -> dict[str, Any]:
    if hashlib.sha256(body).hexdigest() != digest_hex(digest):
        raise ValueError("registry object content digest mismatch")
    return json.loads(body)


def digest_hex(digest: str) -> str:
    algorithm, separator, value = digest.partition(":")
    if separator != ":" or algorithm != "sha256" or len(value) != 64:
        raise ValueError(f"unsupported digest: {digest}")
    int(value, 16)
    return value


def main() -> None:
    args = parse_args()
    expected_root = f"sha256:{digest_hex(args.digest)}"
    root_body = crane_bytes(
        args.crane, "manifest", f"{args.image}@{expected_root}"
    )
    root = parse_verified_json(root_body, expected_root)
    if (
        root.get("schemaVersion") != 2
        or root.get("mediaType") != "application/vnd.oci.image.index.v1+json"
    ):
        raise ValueError("BuildKit result is not an OCI v1 image index")
    manifests = root.get("manifests")
    if not isinstance(manifests, list):
        raise ValueError("BuildKit result is not an OCI image index")

    runnable_digests = {
        descriptor.get("digest")
        for descriptor in manifests
        if descriptor.get("platform", {}).get("os") != "unknown"
    }
    if not runnable_digests or None in runnable_digests:
        raise ValueError("OCI index has no valid runnable manifest")

    subject_predicates: dict[str, set[str]] = {
        digest: set() for digest in runnable_digests
    }
    blobs: list[tuple[str, bytes]] = []
    for descriptor in manifests:
        annotations = descriptor.get("annotations", {})
        if annotations.get("vnd.docker.reference.type") != "attestation-manifest":
            continue
        target = annotations.get("vnd.docker.reference.digest")
        if target not in runnable_digests:
            raise ValueError("attestation descriptor does not target a runnable manifest")

        attestation_digest = descriptor.get("digest")
        digest_hex(attestation_digest)
        manifest_body = crane_bytes(
            args.crane, "manifest", f"{args.image}@{attestation_digest}"
        )
        manifest = parse_verified_json(manifest_body, attestation_digest)
        if (
            manifest.get("schemaVersion") != 2
            or manifest.get("mediaType")
            != "application/vnd.oci.image.manifest.v1+json"
        ):
            raise ValueError("attestation object is not an OCI v1 manifest")
        if manifest.get("artifactType") != ATTESTATION_ARTIFACT_TYPE:
            raise ValueError("attestation manifest does not use the OCI artifact type")
        if manifest.get("subject", {}).get("digest") != target:
            raise ValueError("attestation manifest subject digest mismatch")
        platform = descriptor.get("platform", {})
        if platform.get("os") != "unknown" or platform.get("architecture") != "unknown":
            raise ValueError("attestation descriptor is not unknown/unknown")

        layers = manifest.get("layers")
        if not isinstance(layers, list) or not layers:
            raise ValueError("attestation manifest has no layers")
        for layer in layers:
            if layer.get("mediaType") != IN_TOTO_MEDIA_TYPE:
                raise ValueError("attestation layer is not a full in-toto JSON blob")
            layer_digest = layer.get("digest")
            layer_hex = digest_hex(layer_digest)
            body = crane_bytes(
                args.crane, "blob", f"{args.image}@{layer_digest}"
            )
            if hashlib.sha256(body).hexdigest() != layer_hex:
                raise ValueError("attestation blob content digest mismatch")
            statement = json.loads(body)
            if statement.get("_type") != "https://in-toto.io/Statement/v1":
                raise ValueError("attestation blob is not an in-toto v1 statement")
            predicate_type = statement.get("predicateType")
            if layer.get("annotations", {}).get("in-toto.io/predicate-type") != predicate_type:
                raise ValueError("predicate type annotation mismatch")
            subjects = statement.get("subject")
            if not isinstance(subjects, list) or not any(
                subject.get("digest", {}).get("sha256") == digest_hex(target)
                for subject in subjects
            ):
                raise ValueError("in-toto statement does not bind the target digest")
            predicate = statement.get("predicate")
            if not isinstance(predicate, dict) or not predicate:
                raise ValueError("in-toto statement has no complete predicate object")
            if predicate_type == SBOM_PREDICATE and not predicate.get("spdxVersion"):
                raise ValueError("native SBOM predicate has no SPDX version")
            if predicate_type == "https://slsa.dev/provenance/v1" and (
                not isinstance(predicate.get("buildDefinition"), dict)
                or not isinstance(predicate.get("runDetails"), dict)
            ):
                raise ValueError("native SLSA v1 predicate is incomplete")
            if predicate_type == "https://slsa.dev/provenance/v0.2" and (
                not predicate.get("buildType")
                or not isinstance(predicate.get("builder"), dict)
            ):
                raise ValueError("native SLSA v0.2 predicate is incomplete")

            subject_predicates[target].add(predicate_type)
            blobs.append(
                (f"{digest_hex(target)[:12]}-{layer_hex[:12]}.intoto.json", body)
            )

    for target, predicates in subject_predicates.items():
        if SBOM_PREDICATE not in predicates:
            raise ValueError(f"native BuildKit SBOM is missing for {target}")
        if not predicates.intersection(SLSA_PREDICATES):
            raise ValueError(f"native BuildKit SLSA provenance is missing for {target}")

    args.output_dir.mkdir(mode=0o700, parents=True, exist_ok=False)
    for filename, body in blobs:
        (args.output_dir / filename).write_bytes(body)
    verification = {
        "rootDigest": expected_root,
        "subjects": {
            target: sorted(predicates)
            for target, predicates in sorted(subject_predicates.items())
        },
        "verified": True,
    }
    (args.output_dir / "verification.json").write_text(
        json.dumps(verification, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )


if __name__ == "__main__":
    main()

File: scripts/verify_buildkit_attestations.ts
#!/usr/bin/env -S node --import tsx
/** Fail closed unless BuildKit published complete SBOM and SLSA in-toto blobs. */

import { createHash } from "node:crypto";
import { mkdirSync, writeFileSync } from "node:fs";
import { spawnSync } from "node:child_process";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

export const IN_TOTO_MEDIA_TYPE = "application/vnd.in-toto+json";
export const ATTESTATION_ARTIFACT_TYPE = "application/vnd.docker.attestation.manifest.v1+json";
export const SBOM_PREDICATE = "https://spdx.dev/Document";
export const SLSA_PREDICATES = new Set([
  "https://slsa.dev/provenance/v0.2",
  "https://slsa.dev/provenance/v1",
]);

type JsonObject = Record<string, unknown>;
export type CraneRunner = (arguments_: string[]) => Buffer;

function object(value: unknown, message: string): JsonObject {
  if (value === null || Array.isArray(value) || typeof value !== "object") throw new Error(message);
  return value as JsonObject;
}

function array(value: unknown, message: string): unknown[] {
  if (!Array.isArray(value)) throw new Error(message);
  return value;
}

function sha256(body: Buffer): string {
  return createHash("sha256").update(body).digest("hex");
}

export function digestHex(digest: unknown): string {
  if (typeof digest !== "string" || !/^sha256:[0-9a-f]{64}$/.test(digest)) {
    throw new Error(`unsupported digest: ${String(digest)}`);
  }
  return digest.slice("sha256:".length);
}

function parseVerifiedJson(body: Buffer, digest: string): JsonObject {
  if (sha256(body) !== digestHex(digest)) throw new Error("registry object content digest mismatch");
  return object(JSON.parse(body.toString("utf8")), "registry object is not a JSON object");
}

export function createCraneRunner(crane: string): CraneRunner {
  return (arguments_) => {
    const result = spawnSync(crane, arguments_, {
      encoding: "buffer",
      maxBuffer: 128 * 1024 * 1024,
      shell: false,
      stdio: ["ignore", "pipe", "pipe"],
    });
    if (result.error) throw result.error;
    if (result.status !== 0) {
      throw new Error(`crane exited unsuccessfully (${result.status ?? result.signal ?? "unknown"})`);
    }
    return result.stdout;
  };
}

function stringField(container: JsonObject, name: string, message: string): string {
  const value = container[name];
  if (typeof value !== "string") throw new Error(message);
  return value;
}

export interface VerificationOptions {
  crane: string;
  image: string;
  digest: string;
  outputDirectory: string;
  runner?: CraneRunner;
}

export function verifyBuildkitAttestations(options: VerificationOptions): JsonObject {
  const rootDigest = `sha256:${digestHex(options.digest)}`;
  const runner = options.runner ?? createCraneRunner(options.crane);
  const rootBody = runner(["manifest", `${options.image}@${rootDigest}`]);
  const root = parseVerifiedJson(rootBody, rootDigest);
  if (
    root.schemaVersion !== 2 ||
    root.mediaType !== "application/vnd.oci.image.index.v1+json"
  ) throw new Error("BuildKit result is not an OCI v1 image index");
  const descriptors = array(root.manifests, "BuildKit result is not an OCI image index").map((entry) =>
    object(entry, "OCI descriptor is not an object"),
  );
  const runnableDigests = new Set<string>();
  for (const descriptor of descriptors) {
    const platform = object(descriptor.platform, "OCI descriptor has no platform");
    if (platform.os !== "unknown") runnableDigests.add(`sha256:${digestHex(descriptor.digest)}`);
  }
  if (runnableDigests.size === 0) throw new Error("OCI index has no valid runnable manifest");

  const predicates = new Map([...runnableDigests].map((digest) => [digest, new Set<string>()]));
  const blobs: Array<{ filename: string; body: Buffer }> = [];
  for (const descriptor of descriptors) {
    const annotations = object(descriptor.annotations ?? {}, "descriptor annotations are invalid");
    if (annotations["vnd.docker.reference.type"] !== "attestation-manifest") continue;
    const target = stringField(
      annotations,
      "vnd.docker.reference.digest",
      "attestation descriptor has no target digest",
    );
    if (!runnableDigests.has(target)) {
      throw new Error("attestation descriptor does not target a runnable manifest");
    }
    const platform = object(descriptor.platform, "attestation descriptor has no platform");
    if (platform.os !== "unknown" || platform.architecture !== "unknown") {
      throw new Error("attestation descriptor is not unknown/unknown");
    }
    const attestationDigest = `sha256:${digestHex(descriptor.digest)}`;
    const manifestBody = runner(["manifest", `${options.image}@${attestationDigest}`]);
    const manifest = parseVerifiedJson(manifestBody, attestationDigest);
    if (
      manifest.schemaVersion !== 2 ||
      manifest.mediaType !== "application/vnd.oci.image.manifest.v1+json"
    ) throw new Error("attestation object is not an OCI v1 manifest");
    if (manifest.artifactType !== ATTESTATION_ARTIFACT_TYPE) {
      throw new Error("attestation manifest does not use the OCI artifact type");
    }
    const subject = object(manifest.subject, "attestation manifest has no subject");
    if (subject.digest !== target) throw new Error("attestation manifest subject digest mismatch");
    const layers = array(manifest.layers, "attestation manifest has no layers");
    if (layers.length === 0) throw new Error("attestation manifest has no layers");
    for (const entry of layers) {
      const layer = object(entry, "attestation layer is invalid");
      if (layer.mediaType !== IN_TOTO_MEDIA_TYPE) {
        throw new Error("attestation layer is not a full in-toto JSON blob");
      }
      const layerDigest = `sha256:${digestHex(layer.digest)}`;
      const body = runner(["blob", `${options.image}@${layerDigest}`]);
      if (sha256(body) !== digestHex(layerDigest)) {
        throw new Error("attestation blob content digest mismatch");
      }
      const statement = object(JSON.parse(body.toString("utf8")), "attestation blob is not JSON");
      if (statement._type !== "https://in-toto.io/Statement/v1") {
        throw new Error("attestation blob is not an in-toto v1 statement");
      }
      const predicateType = stringField(statement, "predicateType", "attestation has no predicate type");
      const layerAnnotations = object(layer.annotations, "attestation layer has no annotations");
      if (layerAnnotations["in-toto.io/predicate-type"] !== predicateType) {
        throw new Error("predicate type annotation mismatch");
      }
      const subjects = array(statement.subject, "in-toto statement has no subjects");
      if (
        !subjects.some((entry) => {
          const candidate = object(entry, "in-toto subject is invalid");
          const digest = object(candidate.digest, "in-toto subject has no digest");
          return digest.sha256 === digestHex(target);
        })
      ) throw new Error("in-toto statement does not bind the target digest");
      const predicate = object(statement.predicate, "in-toto statement has no complete predicate object");
      if (Object.keys(predicate).length === 0) throw new Error("in-toto statement has no complete predicate object");
      if (predicateType === SBOM_PREDICATE && typeof predicate.spdxVersion !== "string") {
        throw new Error("native SBOM predicate has no SPDX version");
      }
      if (
        predicateType === "https://slsa.dev/provenance/v1" &&
        (predicate.buildDefinition === null || typeof predicate.buildDefinition !== "object" ||
          predicate.runDetails === null || typeof predicate.runDetails !== "object")
      ) throw new Error("native SLSA v1 predicate is incomplete");
      if (
        predicateType === "https://slsa.dev/provenance/v0.2" &&
        (typeof predicate.buildType !== "string" || predicate.builder === null || typeof predicate.builder !== "object")
      ) throw new Error("native SLSA v0.2 predicate is incomplete");
      predicates.get(target)?.add(predicateType);
      blobs.push({
        filename: `${digestHex(target).slice(0, 12)}-${digestHex(layerDigest).slice(0, 12)}.intoto.json`,
        body,
      });
    }
  }
  for (const [target, types] of predicates) {
    if (!types.has(SBOM_PREDICATE)) throw new Error(`native BuildKit SBOM is missing for ${target}`);
    if (![...SLSA_PREDICATES].some((type) => types.has(type))) {
      throw new Error(`native BuildKit SLSA provenance is missing for ${target}`);
    }
  }

  mkdirSync(options.outputDirectory, { recursive: false, mode: 0o700 });
  for (const { filename, body } of blobs) writeFileSync(join(options.outputDirectory, filename), body, { mode: 0o600 });
  const subjects = Object.fromEntries(
    [...predicates.entries()]
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([target, types]) => [target, [...types].sort()]),
  );
  const verification = { rootDigest, subjects, verified: true };
  writeFileSync(
    join(options.outputDirectory, "verification.json"),
    `${JSON.stringify(verification, null, 2)}\n`,
    { mode: 0o600 },
  );
  return verification;
}

function parseArguments(argv: string[]): VerificationOptions {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    const name = argv[index];
    const value = argv[index + 1];
    if (!name?.startsWith("--") || value === undefined) throw new Error(`invalid argument: ${name ?? "<missing>"}`);
    values.set(name, value);
  }
  const crane = values.get("--crane");
  const image = values.get("--image");
  const digest = values.get("--digest");
  const outputDirectory = values.get("--output-dir");
  if (!crane || !image || !digest || !outputDirectory) {
    throw new Error("required: --crane PATH --image IMAGE --digest sha256:... --output-dir PATH");
  }
  return { crane, image, digest, outputDirectory };
}

export function main(argv = process.argv.slice(2)): void {
  verifyBuildkitAttestations(parseArguments(argv));
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  try {
    main();
  } catch (error) {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
    process.exitCode = 1;
  }
}
